### PR TITLE
annotations experiment

### PR DESCRIPTION
**Pull request by sh-rp**
### Description

This PR has an example of how we could leverage python annotations to have a more universal and more readable format of our schema.

Check out test_pipeline.py on what the interface for the user would be in this example. It actually already works. Basically all you need to to is to put our annotations on class vars, and you can use any new or pre-existing class as a basis for our schema. Typehints that are unknown to us will be ignored (as defined per "Annotated" PEP)

#### Notes and Considerations:
* Not all hints and data_types are support here, this is a prototype, but one that is easy to extend
* There is a meta data attribute "__table__" on the class, we can use this to set table level hints, currently the dlt core is not set up to support this via the columns attribute, but I think it should be quite easy to do.

### CODE REVIEW

File: test_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    p = dlt.pipeline("my_pipe", destination="duckdb", full_refresh=True)
+
+    data = [{
+        "id": "my_id"
+    }]
+
+    # run simple pipeline and see wether schema was used
+    load_info = p.run(data, columns=Items, table_name="blah")
+    print(load_info)
+    print(p.default_schema.to_pretty_yaml())
```
Comment (sh-rp) [inline test_pipeline.py:35]: The items example will produce this segment in the final schema:
```
  blah:
    columns:
      id:
        data_type: text
        primary_key: true
        unique: true
      name:
        data_type: text
        nullable: true
        x-classifiers:
        - pii.name
      email:
        data_type: text
        nullable: true
        unique: true
        x-classifiers:
        - pii.email
      likes_herring:
        data_type: bool
        x-classifiers:
        - pii.food_preference
      _dlt_load_id:
        data_type: text
        nullable: false
      _dlt_id:
        data_type: text
        nullable: false
        unique: true
    write_disposition: append
```

File: dlt/common/schema/annotations.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+class WriteDisposition:
+    value: TWriteDisposition
+
+
+#
+# Converters
+#
+
+TypeMap: Dict[Any, TDataType] = {
+    str: "text",
```
Comment (rudolfix) [inline dlt/common/schema/annotations.py:66]: we have py_type_to_sc_type. look at the type_helpers.py. tons of edge cases are handled there when converting types

File: dlt/common/schema/annotations.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    float: "double",
+    date: "date",
+    datetime: "time",
+    bool: "bool",
+    Decimal: "decimal",
+}
+
+
+def unwrap(t: Type[Any]) -> Tuple[Any, List[Any]]:
+    """Returns python type info and wrapped types if this was annotated type"""
```
Comment (rudolfix) [inline dlt/common/schema/annotations.py:77]: look at typing.py in common, I think I have similar function

File: test_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+from dlt import annotations as a
+from dlt.common import json
+from typing_extensions import Annotated, Never, Optional
+
+
+class Items:
+
+    # metadata for the table, currently not picked up by the pipeline
+    __table__: Annotated[Never, a.TableName("my_items"), a.WriteDisposition("merge")]
```
Comment (rudolfix) [inline test_pipeline.py:11]: hmmm there must be a better way! ie.
load_info = p.run(data, columns=Annotated[Items, a.TableName("my_items"), a.WriteDisposition("merge")] , table_name="blah")

if items are not annotated then we go to default: Items as name, append and write dispositions

Comment (rudolfix) [reply test_pipeline.py:11]: also using `columns` for the above is a kind of abuse... let's sync on table level hints. they may require to change our core library. ie. to make resource a generic that takes model as T. also see: https://github.com/dlt-hub/dlt/issues/753

Comment (rudolfix) [reply test_pipeline.py:11]: or we add "model" to resource definition. but then we are going into a big overhaul of our schema system where relational and python schemas are different.

File: test_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    # primary keys
+    id: Annotated[str, a.PrimaryKey, a.Unique]
+
+    # additional columns
+    name: Annotated[Optional[str], a.Classifiers(["pii.name"])]
```
Comment (rudolfix) [inline test_pipeline.py:17]: I'd probably generate literals for those

**Review by rudolfix — CHANGES_REQUESTED** [review]
this is soo cool

File: dlt/common/schema/annotations.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    if get_origin(t) is Annotated:
+        args = get_args(t)
+        return args[0], list(args[1:])
+    return t, []
+
+
+def to_full_type(t: Type[Any]) -> TColumnSchema:
+    result: TColumnSchema = {}
+    if get_origin(t) is Union:
+        for arg in get_args(t):
```
[the remark this answers is not in the document]
Comment (sh-rp) [reply dlt/common/schema/annotations.py:87]: we can only have one type in the schema, so either we have a default way of resolving if there are multiple types or we throw an error.

[the remark this answers is not in the document]
Comment (sh-rp) [reply dlt/common/schema/annotations.py:87]: Or types of int and string will produce a string, but I think that is taking it to far for now.

File: test_pipeline.py (hunk 5 above)
Comment (sh-rp) [reply test_pipeline.py:17]: you mean for the classifiers?

File: dlt/common/schema/annotations.py (hunk 2 above)
Comment (sh-rp) [reply dlt/common/schema/annotations.py:66]: ah yes, great :) I was wondering if we had something like this somewhere.

File: dlt/common/schema/annotations.py (hunk 3 above)
Comment (sh-rp) [reply dlt/common/schema/annotations.py:77]: yes, i extracted the essential part.

File: test_pipeline.py (hunk 4 above)
Comment (sh-rp) [reply test_pipeline.py]: i changed it to accept annotated tables now, so this would work.

**Pull request closed without merging**